src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -805,6 +805,162 @@ case class TableDistinct(child: TableIR) extends TableIR {
   }
 }
 
+case class TableKeyByAndAggregate(
```

Comment:
Can you add an optimization rule that turns `AggregateByKey(KeyBy())` into `KeyByAndAggregate()`.

Reply:
And maybe something that turns a `KeyByAndAggregate(x)` into `AggregateByKey(x)` when `x` is already keyed by `newKey`.

Reply:
yep, good call

Reply:
added.